tools/logger/filter.c

```diff
@@ -0,0 +1,316 @@
+#include "filter.h"
+
+#define COMPONENTS_SEPARATOR ','
+
+/** map between log level given by user and enum value */
+static const struct {
+	const char name[16];
+	int32_t log_level;
+} log_level_dict[] = {
+	{"verbose",	LOG_LEVEL_VERBOSE},
+	{"debug",	LOG_LEVEL_DEBUG},
+	{"info",	LOG_LEVEL_INFO},
+	{"warning",	LOG_LEVEL_WARNING},
+	{"error",	LOG_LEVEL_ERROR},
+	{"critical",	LOG_LEVEL_CRITICAL},
+	{"v",		LOG_LEVEL_VERBOSE},
+	{"d",		LOG_LEVEL_DEBUG},
+	{"i",		LOG_LEVEL_INFO},
+	{"w",		LOG_LEVEL_WARNING},
+	{"e",		LOG_LEVEL_ERROR},
+	{"c",		LOG_LEVEL_CRITICAL},
+};
+
+struct filter_element {
+	struct list_item list;
+	struct sof_ipc_dma_trace_filter_elem ipc_elem;
+};
+
+static char *ltrim(char *s)
+{
+	while (isspace(*s))
+		s++;
+	return s;
+}
+
+static char *rtrim(char *s)
+{
+	char *ptr = s + strlen(s) - 1;
+
+	while (ptr >= s && isspace(*ptr)) {
+		*ptr = '\0';
+		--ptr;
+	}
+	return s;
+}
+
+static char *trim(char *s)
+{
+	return ltrim(rtrim(s));
+}
+
+/**
+ * Search for uuid entry with given component name in given uids dictionary
+ * @param uids_dict dictionary to search in
+ * @param name of uuid entry
+ * @return pointer to sof_uuid_entry with given name
+ */
+static struct sof_uuid_entry *
+get_uuid_by_name(const struct snd_sof_uids_header *uids_dict,
+		 const char *name)
+{
+	uintptr_t beg = (uintptr_t)uids_dict + uids_dict->data_offset;
+	uintptr_t end = beg + uids_dict->data_length;
+	struct sof_uuid_entry *ptr = (struct sof_uuid_entry *)beg;
+
+	while ((uintptr_t)ptr < end) {
+		if (strcmp(name, ptr->name) == 0)
+			return ptr;
+		++ptr;
+	}
+	return NULL;
+}
+
+/**
+ * Parse log level name (from log_level_dict) to enum value.
+ * Take care about possible whitespace at the begin.
+ * @param value_start pointer to the begin of range to search
+ * @return enum value for given log level, or -1 for invalid value
+ */
+static int filter_parse_log_level(const char *value_start)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(log_level_dict); ++i) {
+		if (strstr(log_level_dict[i].name, value_start))
+			return log_level_dict[i].log_level;
+	}
+	return -1;
+}
+
+/**
+ * Parse component definition from input_str.
+ *
+ * Possible input_str formats:
+ *   `name pipe_id.comp_id`
+ *   `name pipe_id.*`
+ *   `name *`
+ *   `name`
+ *   `*`
+ * Whitespace is possible at the begin, end and after `name`.
+ * `name` must refer to values from given UUID dictionary,
+ *        (so name comes from DECLARE_SOF_UUID macro usage)
+
+ * @param uids_dict dictionary with list of possible `name` values
+ * @param input_str formatted component definition
+ * @param out element where component definition should be saved
+ */
+static int filter_parse_component(const struct snd_sof_uids_header *uids_dict,
+				  char *input_str,
+				  struct sof_ipc_dma_trace_filter_elem *out)
+{
+	struct sof_uuid_entry *uuid_entry;
+	char comp_name[UUID_NAME_MAX_LEN];
+	char scan_format_string[32];
+	const char *instance_info;
+	static const struct sof_uuid sof_pipe_uuid = {
```

Comment:
We should not be hard coding these system UUIDs. I think we probably need a shared header ?

Reply:
I have try to move pipe uuid A/B/C/D values to headers in user/ folder, but this solution doesn't look nice. So I just remove this part.